### DISCUSSION TITLE
M32 is it supported? Trying to use M32 to call progs as sub routine.

### DISCUSSION THREAD

**Discussion by TRX0011** [mriscoc/Ender3V2S1#1113] [Q&A]
Hello everyone,

I am trying to use the M32 command to call within a program another gcode file as a sub routine and then to continue execute the rest of the code. I want to use this method instead of M808, in order to keep the code smaller, tidy and able to make corrections easier than having a huge file with multiple M808 loops.

Tried to use the M32 command with no luck. What I did:
Created a master gcode file (FILE1.gcode) in root.
The 2nd file (subroutine, FILE2.gcode, also in root) was called inside file1
I had this error:

echo:Unknown command: "M32 P!FILE2~1.GCO"

Tried with long filenames, in root, in sub directory, with /,  nothing

In Configuration.h, in SD CARD section, is uncomment:
#define SDSUPPORT  // Ender Configs

What am I doing wrong? 
Can anyone please post some code, as an example? 

Is it needed  to give any special command at the end of the file that was called, in order to return to the master program?

Thank you,

Tom

**Comment by mriscoc** [comment]
SD_PROCEDURE_DEPTH (M32) is disabled in current precompiled versions; you can enable it in Configuration_adv.h file and compile a custom build or request a version with SD_PROCEDURE_DEPTH enabled.

https://github.com/mriscoc/Special_Configurations#how-to-request-a-custom-build

**Reply by TRX0011** [reply]
The only SD-PROCEDURE I found was in the CARDREADER.h file and found this:

 //
  // Procedure calls to other files
  //
  #if HAS_MEDIA_SUBCALLS
    static uint8_t file_subcall_ctr;
    static uint32_t filespos[SD_PROCEDURE_DEPTH];
    static char proc_filenames[SD_PROCEDURE_DEPTH][MAXPATHNAMELENGTH];
  #endif

**_Should I changed it to this?_**

 
  #if HAS_MEDIA_SUBCALLS
    **#define SD_PROCEDURE_DEPTH 2**
    static uint8_t file_subcall_ctr;
    static uint32_t filespos[SD_PROCEDURE_DEPTH];
    static char proc_filenames[SD_PROCEDURE_DEPTH][MAXPATHNAMELENGTH];
  #endif

**Reply by TRX0011** [reply]
My mistake for not reading correctly your answer. You were right, by uncommented the #define SD_PROCEDURE_DEPTH 1 in Configuration_adv.h file, I was able to use M32 command. But unsuccessfully.

I have created a test .gcode file, which calls (M32) two files to make some moves. What I got, is the main program runs from beginning and ends, ignoring the M32 subroutines. When it ends, THEN runs the called program #1, called program#2... 
It behaves like flushing the buffer, first executing all commands from the main program, then the others, like FIFO.
Is the # at the end of the command, indicates to stop buffering and wait for the called program to end, before proceed to the next line?

Tried also #define SD_PROCEDURE_DEPTH 2 with no luck.

What I have to do in order to work properly?

Does anyone has a sample code to show how to use M32 inside a file?

Here are my test files:

  MAIN FILE.......
; test M32 
G28 X
G0 X100
G0 X30
M117 Calling File 2
M118 Calling File 2
M32 P !/FILE2~1.gco#
G4 S3
G0 X100
M32 P!/FILE3~1.gco#
G4 S3
G28 X

FILE2...
G0 Y100
G0 Y0
M400

FILE3..
G0 Z100
G0 Z30
M400

**Comment by mriscoc** [comment]
Respect space chars
https://marlinfw.org/docs/gcode/M032.html
